e2e/pos_test.go

```diff
@@ -121,3 +139,108 @@ func TestPoS_Unstake(t *testing.T) {
 	assert.Nil(t, validator, "account should have left from validator set, but still belongs to it")
 	assert.Len(t, snapshot.Validators, numGenesisValidators-1)
 }
+
+func TestPoS_UnstakeAttack(t *testing.T) {
+	// Predefined values
+	unstakingContractAddr := types.StringToAddress(system.UnstakingAddress)
+	numGenesisValidators := IBFTMinNodes + 1
+	gasPrice := big.NewInt(10000)
+	receiptArr := make([]*web3.Receipt, numGenesisValidators)
+	defaultBalance := currency.EthToWei(10)
+
+	// Set up the manager
+	ibftManager := framework.NewIBFTServersManager(t, numGenesisValidators, IBFTDirPrefix, func(i int, config *framework.TestServerConfig) {
+		config.PremineValidatorBalance(defaultBalance, defaultBalance)
+		config.SetSeal(true)
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	ibftManager.StartServers(ctx)
+	srv := ibftManager.GetServer(0)
+
+	// Get key of last node
+	referenceSrv := ibftManager.GetServer(IBFTMinNodes)
+	referenceKey, err := referenceSrv.Config.PrivateKey()
+	assert.NoError(t, err)
+	referenceAddr := crypto.PubKeyToAddress(&referenceKey.PublicKey)
+
+	// Check the validator is in validator set
+	validator, _ := findValidatorInSet(t, srv, referenceAddr, 0)
+	if validator == nil {
+		t.Fatalf("account should be genesis validator, but isn't")
+	}
+
+	// Set up the goroutine
+	sendFunc := func(wg *sync.WaitGroup, t *testing.T, index int) {
+		txn := &framework.PreparedTransaction{
+			From:     referenceAddr,
+			To:       &unstakingContractAddr,
+			GasPrice: gasPrice,
+			Gas:      1000000,
+			Value:    currency.EthToWei(0),
+		}
+
+		ctx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		receipt, err := srv.SendRawTx(ctx, txn, referenceKey)
+		assert.NoError(t, err)
+		assert.NotNil(t, receipt)
+
+		receiptArr[index] = receipt
+
+		wg.Done()
+	}
+
+	// Add pending unstake transactions to the txpool
+	numTransactions := 5
+
+	// Send out multiple unstake transactions
+	var wg sync.WaitGroup
+
+	// WaitGroup will wait for numTransactions go routines to finish
+	wg.Add(numTransactions)
+	for i := 0; i < numTransactions; i++ {
+		go sendFunc(&wg, t, i)
+	}
+
+	wg.Wait()
+
+	fee := new(big.Int).Mul(
+		big.NewInt(int64(receiptArr[0].GasUsed)),
+		big.NewInt(gasPrice.Int64()),
+	)
+
+	// Check to see if balances are valid
+
+	// Returned stake
+	expectedBalance := big.NewInt(0).Add(defaultBalance, defaultBalance)
+
+	// Subtract tx fees
+	expectedBalance = big.NewInt(0).Sub(
+		expectedBalance,
+		fee,
+	)
+
+	actualStakedBalance := getStakedBalance(referenceAddr, srv.JSONRPC(), t)
+	actualBalance := getAccountBalance(referenceAddr, srv.JSONRPC(), t)
```

Comment:
Nitpick warning:
You can do this, to prevent instantiation of multiple JSONRPC clients:
```suggestion
	client := srv.JSONRPC()
	actualStakedBalance := getStakedBalance(referenceAddr, client, t)
	actualBalance := getAccountBalance(referenceAddr, client, t)
```

Reply:
Good catch.

Fixed 👍